File: 0051-n-queens/0051-n-queens.cpp
class Solution {
public:
    vector<vector<string>> res;
    unordered_set<int> colSt;
    unordered_set<int> diagSt;
    unordered_set<int> antiDiagSt;
    
    void placeQueens(int n, int row, vector<string> &board){
        if(row >= n){
            res.push_back(board);
            return ;
        }
        for(int col = 0; col<n; col++){
            int diagonal = row + col;
            int antiDiag = row - col;
            if(colSt.find(col) != colSt.end() 
            || diagSt.find(diagonal) != diagSt.end() 
            || antiDiagSt.find(antiDiag) != antiDiagSt.end()){
                continue;
            }
            colSt.insert(col);
            diagSt.insert(diagonal);
            antiDiagSt.insert(antiDiag);
            board[row][col] = 'Q';
            placeQueens(n, row+1, board);
            board[row][col] = '.';
            colSt.erase(col);
            diagSt.erase(diagonal);
            antiDiagSt.erase(antiDiag);
        }
    }
    vector<vector<string>> solveNQueens(int n) {
        vector<string> board(n, string(n, '.'));
        
        placeQueens(n, 0, board);
        return res;

    }
};

File: 0051-n-queens/0051-n-queens.java
class Solution {
    List<List<String>> ans = new ArrayList<>();
    public void placeQueens(List<String> board, int row, int n, HashSet<Integer> colSt, HashSet<Integer> diagSt, HashSet<Integer> antiDiagSt){
        if(row >= n){
            ans.add(new ArrayList<>(board));
            return ;
        }
        for(int col = 0; col < n; col++){
            int diagonal = row-col;
            int antiDiag = row+col;
            if(colSt.contains(col) || diagSt.contains(diagonal) || antiDiagSt.contains(antiDiag)){
                    continue;
            }
            colSt.add(col);
            diagSt.add(diagonal);
            antiDiagSt.add(antiDiag);
            StringBuilder sb = new StringBuilder(board.get(row));
            sb.setCharAt(col, 'Q');
            board.set(row, sb.toString());
            placeQueens(board, row+1, n, colSt, diagSt, antiDiagSt);
            sb.setCharAt(col, '.');
            board.set(row, sb.toString());
            colSt.remove(col);
            diagSt.remove(diagonal);
            antiDiagSt.remove(antiDiag);
        }
    }
    public List<List<String>> solveNQueens(int n) {
        List<String> board = new ArrayList<>();

        HashSet<Integer> colSt = new HashSet<>();
        HashSet<Integer> diagSt = new HashSet<>();
        HashSet<Integer> antiDiagSt = new HashSet<>();

       
        for(int i = 0; i<n; i++){
            StringBuilder sb = new StringBuilder();
            for(int j = 0; j<n; j++){
                sb.append('.');
            }
            board.add(sb.toString());
        }
        placeQueens(board, 0, n, colSt, diagSt, antiDiagSt);
        return ans;
    }
}